### PR TITLE
Fixes ended poll voting

### PR DESCRIPTION
**Pull request by ericdecanini**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Fixes polls being votable after being ended

## Motivation and context

Closes https://github.com/vector-im/element-android/issues/5040

The cause of the issue is found in `EventRelationsAggregationProcessor:388`.

When voting, the SDK checks that the eventTimestamp is before the closed time and if it is, the vote response can go through, despite the poll having already been ended (determined by the non-null presence of a closed time).

The fix on the SDK was to simply remove that extra condition. I checked that this shouldn't have any side-effects. I tried to approach this differently, i.e. trying to fix the event timestamp (`event.originServerTs`) set during the poll ending so that it matches the poll closed time but to no avail.

Though I should be able to investigate this with a bit more time if that's the preferrable course of action

I also added an extra layer of protection on the client side to check that the poll is in an ended state.

## Tests

- Set your device system time to 1 hour ahead of current time
- In any timeline, create a new poll
- Immediately end it
- Try to vote on it and see that it cannot be voted

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): Android 10

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt [hunk 1]
```diff
@@ -385,7 +385,7 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
                 }
 
         val closedTime = existingPollSummary?.closedTime
-        if (closedTime != null && eventTimestamp > closedTime) {
+        if (closedTime != null) {
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:388]: is there a case where the poll could have a closed timestamp but still be valid to casting votes? cc @onurays

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/PollItem.kt [hunk 2]
```diff
@@ -73,13 +72,21 @@ abstract class PollItem : AbsMessageItem<PollItem.Holder>() {
...
 
+    private fun onPollItemClick(optionViewState: PollOptionViewState) {
+        val relatedEventId = eventId
+
+        if (isPollActive(optionViewState) && relatedEventId != null) {
+            callback?.onTimelineItemAction(RoomDetailAction.VoteToPoll(relatedEventId, optionViewState.optionId))
+        }
+    }
+
+    private fun isPollActive(optionViewState: PollOptionViewState) = optionViewState !is PollOptionViewState.PollEnded
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/PollItem.kt:88]: There is too much logic here. Better to move that to the controller.
Maybe replace `var pollSent: Boolean = false` which is unused by something like `var canVote: Boolean = false`

**Review by bmarty — COMMENTED** [review]
One remark.

You may have seen it, but this is where the SDK ignore late vote FWIW: https://github.com/vector-im/element-android/blob/main/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt#L388

**Comment by ericdecanini** [comment]
Update: refactored MessageItemFactory's `buildPollItem` method and a few other poll-related classes

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt (hunk 1 above)
Comment (ericdecanini) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:388]: @onurays bump

**Review by ouchadam — APPROVED** [review]
thanks for all the cleaning up :100: 

would be good to double check with @onurays for the original timestamp condition in case there's a scenario we aren't aware of

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt (hunk 1 above)
Comment (onurays) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:401]: I think we should keep this condition. Even if we prevent it previously, there is always a chance to trigger a poll response to an ended poll from another client or via rest api.

File: vector/src/main/java/im/vector/app/features/poll/PollState.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.poll
+
+sealed class PollState {
```
[the remark this answers is not in the document]
Comment (ericdecanini) [reply vector/src/main/java/im/vector/app/features/poll/PollState.kt]: This has never crossed my mind, but I totally agree! Changes pushed

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt [hunk 4]
```diff
@@ -179,42 +191,44 @@ class MessageItemFactory @Inject constructor(
...
-            is MessageVideoContent               -> buildVideoMessageItem(messageContent, informationData, highlight, callback, attributes)
-            is MessageFileContent                -> buildFileMessageItem(messageContent, highlight, attributes)
-            is MessageAudioContent               -> {
+            is MessageEmoteContent -> buildEmoteMessageItem(messageContent, informationData, highlight, callback, attributes)
+            is MessageTextContent -> buildItemForTextContent(messageContent, informationData, highlight, callback, attributes)
+            is MessageImageInfoContent -> buildImageMessageItem(messageContent, informationData, highlight, callback, attributes)
+            is MessageNoticeContent -> buildNoticeMessageItem(messageContent, informationData, highlight, callback, attributes)
+            is MessageVideoContent -> buildVideoMessageItem(messageContent, informationData, highlight, callback, attributes)
+            is MessageFileContent -> buildFileMessageItem(messageContent, highlight, attributes)
+            is MessageAudioContent -> {
```
Comment (onurays) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/MessageItemFactory.kt:203]: The arrows of when conditions should be at the same column. Could you please check your formatter? (We had this problem on a previous Android Studio).

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt [hunk 5]
```diff
@@ -482,46 +482,39 @@ internal class EventRelationsAggregationProcessor @Inject constructor(
...
 
         // we have it
-        val existingPollSummary = existing.pollResponseSummary
+        val existingPollSummary = existingPoll.pollResponseSummary
                 ?: realm.createObject(PollResponseAggregatedSummaryEntity::class.java).also {
-                    existing.pollResponseSummary = it
+                    existingPoll.pollResponseSummary = it
                 }
 
-        if (existingPollSummary.closedTime != null) {
```
Comment (ericdecanini) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:509]: The source of the bug is here. A local echo being processed before the remote echo meant that the remote echo never reaches past this point. `closedTime` thus never gets set from the remote and always turns out to be the current time millis of the local device that ended the poll. If the device has time that's out of sync with unix current time, this bug can occur

cc @onurays

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt (hunk 1 above)
Comment (ericdecanini) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt]: Reverted this change and pushed alternative solution. See below comment for details

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt (hunk 5 above)
Comment (onurays) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/EventRelationsAggregationProcessor.kt:509]: I see, thanks for the clarification.

**Review by onurays — APPROVED** [review]

**Pull request merged**